gtk/src/toga_gtk/hardware/location.py

```diff
@@ -0,0 +1,146 @@
+from enum import IntEnum, auto
+
+import gi
+
+from toga import LatLng
+from toga_gtk.libs import Gio, GLib, GObject
+
+gi.require_version("Geoclue", "2.0")
+from gi.repository import Geoclue  # noqa: E402
+
+
+def toga_location(location):
+    """Convert a ``Geoclue.Location`` into ``OnLocationChangeHandler`` kwargs."""
+    latlng = LatLng(*location.get_properties("latitude", "longitude"))
+    altitude = location.get_property("altitude")
+
+    return {
+        "location": latlng,
+        "altitude": altitude,
+    }
+
+
+class States(IntEnum):
+    STARTING = auto()
+    """Waiting for the Geoclue client to initialise."""
+
+    READY = auto()
+    """Geoclue is ready to connect."""
+
+    MONITORING = auto()
+    """Actively monitoring the location."""
+
+    FAILED = auto()
+    """``Geoclue`` was unable to retrieve the location due to a generic error."""
+
+    DENIED = auto()
+    """Permission to read the location was denied."""
+
+
+class Location(GObject.Object):
+    # Implement as GObject.Object to allow notifying on state changes within the event loop
+    state = GObject.Property(type=int, default=States.STARTING)
+
+    def __init__(self, interface):
+        super().__init__()
+
+        self.interface = interface
+
+        self.geoclue_simple = None
```

Comment:
By way of consistency, we usually try to keep *one* object on the impl named `native`, as the "main" representative of the platform's underlying implementation.  This geoclue client object would seem to be the best candidate here.

Reply:
Done in https://github.com/beeware/toga/pull/2999/commits/0e0ec80c4e03146e3f4678c29852c2ba48017cf8#diff-7003a350ef59818bcbe863c85ad516382a8e5ae5fcb8eb0bb09fe639e5843e60R55